src/compiler/utilitiesPublic.ts

```diff
@@ -956,6 +961,52 @@ export function getModifiers(node: HasModifiers): readonly Modifier[] | undefine
     }
 }
 
+export function getJSDocCommentsAndTags(hostNode: Node, noCache?: boolean): readonly (JSDoc | JSDocTag)[] {

```

Comment:
This function is wrapped in the language service as getCommentHavingNodes, for declarations only.

This function needs at least two kinds of documentation:

1. Why use it instead of getJSDocTags? They both look for multiple relevant comments, but getJSDocTags only returns tags that apply, not the entire comment. Does non-local comment text ever make sense [1]? Are the non-local comments useful in some way besides comments+tags? Would a caller need to know that one tag was part of a local comment and another was part of a non-local one?
2. What do the parameters mean and what preconditions are there for their use? Specifically:
    - Internally, getJSDocCommentsAndTags is called from two places; the first verifies `canHaveJSDoc(hostNode)` and the second passes only declarations. So I think this only works on specific kinds of node, or maybe on any node, but might hurt performance due to deoptimisation.
    - getJSDocTagsNoCache is labelled internal but find-all-refs doesn't find anything. Why wouldn't you want caching?

[1] eg

```js
/**
 * Function documentation, non-local to x, but shouldn't be needed, right?
 * @param x non-local documentation for x, in a tag
 */
function f(/** also local, so maybe this is needed? */x) { }
```

Reply:
1. It should be used when you want documentation for a node, personally, `getJSDocTags` is something I never see myself using, because if I want documentation for something, I want all related documentation, not only documentation that's nested within a tag. `getJSDocTags` would return nothing of use whatsoever for non-parameters/type parameters... I have no idea why you'd use that function, and was surprised it was exposed while this more useful version was not.
3. I experimented with changing this to accept a `Declaration` rather than `Node`, and while TypeDoc is fine with that signature, `getJSDocTypeParameterTagsWorker` isn't (`InferTypeNode` and `JSDocTemplateTag`)
4. Caching - I removed `noCache` and the other `noCache` parameters from functions that call this with that parameter, and found that `getJSDocOverrideTagNoCache` is public, but that looks like a mistake, the other `getJSDoc*Tag` functions all use the cache. The real user of all of this is `getJSDocModifierFlagsNoCache`, which is ultimately used by `__debugModifierFlags` in `enableDebugInfo`... not familiar enough with the compiler to say if these ought to be cleaned up.